Add event entities to homekit_controller

Jc2k opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add `event` entities for homekit_controller doorbell and buttons.

Todo:

* [x] Tests
* [x] Manual tests


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Jc2k committed:
WIP

Jc2k committed:
Add tests for event

Jc2k committed:
Fix triggers and events concurrently

Jc2k committed:
Better entity names

bdraco commented on homeassistant/components/homekit_controller/event.py:
```diff
@@ -0,0 +1,156 @@
+                    )
+                )
+
+        if entities:
+            async_add_entities(entities)
+            return True
+
+        return False
```
```suggestion
        if entities:
            async_add_entities(entities)

```

typed to `None` but returning bool

Jc2k replied:
The inner function is typed to bool, the outer function is typed to none. It only marks a characteristic as processed if the callback returns true.

bdraco commented:
ordered https://www.amazon.com/dp/B0BPHTL7MG?psc=1&ref=ppx_yo2ov_dt_b_product_details for testing. will be here later today

frenck commented on homeassistant/components/homekit_controller/event.py:
```diff
@@ -0,0 +1,156 @@
+from .entity import HomeKitEntity
+
+INPUT_EVENT_VALUES = {
+    InputEventValues.SINGLE_PRESS: "single_press",
+    InputEventValues.DOUBLE_PRESS: "double_press",
+    InputEventValues.LONG_PRESS: "long_press",
```
We need to add these to `strings.json` to they are translated in the UI

frenck commented on homeassistant/components/homekit_controller/event.py:
```diff
@@ -0,0 +1,156 @@
+                    EventEntityDescription(
+                        key=f"{service.accessory.aid}_{service.iid}",
+                        device_class=EventDeviceClass.DOORBELL,
+                        translation_key="doorbell",
```
We can leave this one out, it will fall back to the translation provided by the device class.

```suggestion
```

frenck commented on homeassistant/components/homekit_controller/event.py:
```diff
@@ -0,0 +1,156 @@
+                        key=f"{service.accessory.aid}_{service.iid}",
+                        device_class=EventDeviceClass.DOORBELL,
+                        translation_key="doorbell",
+                        name="Doorbell",
```
Setting both a name and a translation key will not work 😉 We can just remove it... translations all the way :)

```suggestion
```

frenck commented on homeassistant/components/homekit_controller/event.py:
```diff
@@ -0,0 +1,156 @@
+                        EventEntityDescription(
+                            key=f"{service.accessory.aid}_{service.iid}",
+                            device_class=EventDeviceClass.BUTTON,
+                            translation_key="button",
+                            name=f"Button {idx}",
```
This will not work? The name is hardcoded but a translation key is provided. The translation key wins in this case (and could also be omitted to use the device class one). That said, the `{idx}` is thus not used.

Jc2k commented:
We can't currently support translations for entity names because of legacy stuff that needs refactoring. There is a draft PR that enables support for the new entity naming scheme that needs finishing, and then I have some other branches locally to switch entities over to using translations. There's still some weirdness to resolve around entity ids when name is none and translations are on. They don't seem to match the translation and are not very good entity ids full stop. I can get acceptable English ids if I set the name AND translation key. We also have some cases right now where we pick up a name the user chose through their devices ecosystem. Eg they already names in "Bedroom Motion Sensor".
The first naive attempt broke pretty much every test. There's a lot of work to make sure there's no (important) regressions.

Jc2k commented:
IIRC, I think my other PR was basically finished. There were changes requested but there was an agreement that they could be done in the follow up PRs (as the strings would all be moved to strings.json in those PRs). So if translations are a blocker will get that PR rebased are ready to land, and come back to this PR after that.

frenck commented:
> We can't currently support translations for entity names because of legacy stuff that needs refactoring. 

Alright 👍

bdraco commented:
Device showed up later than expected so will test in the morning

Jc2k committed:
Remove names that are wrong

Jc2k committed:
Add translations for state attributes

Jc2k committed:
Fix lint

Jc2k committed:
Consistent strings.json

Jc2k commented:
Successfully tested with a Wemo Stage (in bluetooth mode).

<img width="601" alt="Screenshot 2023-07-25 at 13 37 45" src="https://github.com/home-assistant/core/assets/11544/c03ee22a-a1b1-4499-99d8-0b7257c5a87d">

State attributes are translated. It uses the existing homekit_controller entity name logic for now, so dropped the `name` field on the EventEntityDescription for the time being.

Jc2k committed:
Format

bdraco commented:
<img width="573" alt="Screenshot 2023-07-25 at 10 25 25 AM" src="https://github.com/home-assistant/core/assets/663432/6f30303f-9fea-426b-ac35-b5ed78829cfc">


works great with the test device

bdraco commented:
Just found a keyerror in the log 
```
25 15:25:47 homeassistant homeassistant[536]: 2023-07-25 10:25:47.075 ERROR (MainThread) [homeassistant.components.homekit_controller.connection] Unexpected exception from <bound method HKDevice.async_update of <homeassistant.components.homekit_controller.connection.HKDevice object at 0x7fe92a781050>>
Jul 25 15:25:47 homeassistant homeassistant[536]: Traceback (most recent call last):
Jul 25 15:25:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/helpers/debounce.py", line 116, in _handle_timer_finish
Jul 25 15:25:47 homeassistant homeassistant[536]:     await task
Jul 25 15:25:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/connection.py", line 757, in async_update
Jul 25 15:25:47 homeassistant homeassistant[536]:     self.process_new_events(new_values_dict)
Jul 25 15:25:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/connection.py", line 778, in process_new_events
Jul 25 15:25:47 homeassistant homeassistant[536]:     callback_()
Jul 25 15:25:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/event.py", line 78, in _handle_event
Jul 25 15:25:47 homeassistant homeassistant[536]:     self._trigger_event(INPUT_EVENT_VALUES[self._characteristic.value])
Jul 25 15:25:47 homeassistant homeassistant[536]:                         ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 25 15:25:47 homeassistant homeassistant[536]: KeyError: None
Jul 25 15:26:47 homeassistant homeassistant[536]: 2023-07-25 10:26:47.077 ERROR (MainThread) [homeassistant.components.homekit_controller.connection] Unexpected exception from <bound method HKDevice.async_update of <homeassistant.components.homekit_controller.connection.HKDevice object at 0x7fe92a781050>>
Jul 25 15:26:47 homeassistant homeassistant[536]: Traceback (most recent call last):
Jul 25 15:26:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/helpers/debounce.py", line 116, in _handle_timer_finish
Jul 25 15:26:47 homeassistant homeassistant[536]:     await task
Jul 25 15:26:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/connection.py", line 757, in async_update
Jul 25 15:26:47 homeassistant homeassistant[536]:     self.process_new_events(new_values_dict)
Jul 25 15:26:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/connection.py", line 778, in process_new_events
Jul 25 15:26:47 homeassistant homeassistant[536]:     callback_()
Jul 25 15:26:47 homeassistant homeassistant[536]:   File "/usr/src/homeassistant/homeassistant/components/homekit_controller/event.py", line 78, in _handle_event
Jul 25 15:26:47 homeassistant homeassistant[536]:     self._trigger_event(INPUT_EVENT_VALUES[self._characteristic.value])
Jul 25 15:26:47 homeassistant homeassistant[536]:                         ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 25 15:26:47 homeassistant homeassistant[536]: KeyError: None

```

Jc2k commented:
If it's none it's not been read at all, so surprised it's getting there. But will add a guard to drop.

frenck commented on homeassistant/components/homekit_controller/strings.json:
```diff
@@ -71,6 +71,30 @@
+        "state_attributes": {
+          "event_type": {
+            "state": {
+              "double_press": "Double Press",
+              "long_press": "Long Press",
+              "single_press": "Single Press"
```
```suggestion
              "double_press": "Double press",
              "long_press": "Long press",
              "single_press": "Single press"
```

frenck commented on homeassistant/components/homekit_controller/strings.json:
```diff
@@ -71,6 +71,30 @@
+          }
+        }
+      },
+      "button": {
```
If they are the same, we can just use one and the same translation key and omit one?

Jc2k replied:
When I do finish supporting translations for the name, no I'll need different translation keys. Is there a syntax to refer to the other keys within the same strings.json?

frenck replied:
Yes, there is, check out `light/strings.json`, I know many of them are in there... :)

Jc2k committed:
Update homeassistant/components/homekit_controller/strings.json

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

Jc2k committed:
Fix conflict between device automations.

Old code tried to work around this by treating event entities as
characteristic entities. This broken handling of remotes (dupe
entities).

New code does it the right way and fixes device automation triggers
to not misuse conn.add_listener.

Jc2k committed:
For IP backed devices, the spec allows the event characteristic to
be marked as pollable, but explicitly returns None.

If we do poll an event characteristic we must ignore its value
as no button press etc has happened.

Jc2k committed:
Don't dupe translations.

Jc2k commented:
This should fix the traceback @bdraco saw, avoid dupes in the new translations and fix a conflict with the old device triggers the proper way.

bdraco reviewed (approved):
Tested, working as expected

Jc2k commented:
Thanks! 🎉

frenck reviewed (approved):
Thanks, @Jc2k 👍 

../Frenck

MartinHjelmare commented on homeassistant/components/homekit_controller/event.py:
```diff
@@ -0,0 +1,160 @@
+
+        self.entity_description = entity_description
+
+        # An INPUT_EVENT may support single_press, long_press and double_press. All are optional. So we have to
```
Please break long strings and comments around 88 characters per line.
